XARCH: Eliminate redundant mov to/from mem.

DeepakRajendrakumaran opened the pull request:
Optimize redundant mov instructions like:

IN002f: 000141 mov      qword ptr [V08 rsp+78H], rax
IN0030: 000146 mov      rax, qword ptr [V08 rsp+78H]
see comment ([link](https://github.com/dotnet/runtime/issues/12984#issuecomment-888525339))

This commit will essentially mirror what we do for Register->Register movs but with slight modifications to customize it for mem<->reg mov.
There are a couple of further optimizations to be revisited for both this and the existing IsRedundantMov() implementation(These are already mentioned as comments in existing impl)

1. Neither will work for sse on targets with avx encoding since it cannot be determined the current way it's implemented if the upper bits have been modified.
2. Currently, we check if both instructions(curr and predecessor) are equivalent. We can extend it for some other cases like aligned/unaligned(movups/movaps for eg.)

An example code snippet to reproduce this
`
    [MethodImpl(MethodImplOptions.AggressiveOptimization)]

    static unsafe double fmaTest4()

    {
        vec b;
 
       var a = Vector256.AsDouble(Vector256.Create(3f));

        var c = Vector256.AsDouble(Vector256.Create(2f));

        var d = Vector256.AsDouble(Vector256.Create(3f));

        c = Fma.MultiplyAdd(Avx.LoadVector256((double*)&a), Avx.LoadVector256((double*)&b), Avx.LoadVector256((double*)&c));

        return Avx.Add(Avx.LoadVector256((double*)&c), Avx.LoadVector256((double*)&d)).ToScalar();

    }`

DeepakRajendrakumaran committed:
XARCH: Eliminate redundant mov to/from mem.

tannergooding commented:
CC. @JulieLeeMSFT, @dotnet/jit-contrib

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -7055,12 +7055,170 @@ void emitter::emitIns_S(instruction ins, emitAttr attr, int varx, int offs)
+//         vmovapd  xmmword ptr [V01 rbp-20H], xmm0  # <-- last instruction
+//         vmovapd  xmmword ptr [V01 rbp-20H], xmm0  # <-- current instruction can be omitted.
+//
+//    2. Opposite Move as that of last instruction emitted.
```
Isn't this case `*_R_S` rather than `*_S_R`?

DeepakRajendrakumaran replied:
It covers both R_S and S_R

tannergooding replied:
Maybe we could call this `IsRedundantStackMov` or similar?

Someone on the JIT team may have a better suggestion or convention we could use here.

DeepakRajendrakumaran replied:
This handles the following cases
1. mov reg mem
   mov reg mem // identical instructions mem ->reg
2. mov mem reg
   mov mem reg // identical instructions reg ->mem
3. mov mem reg
   mov reg mem  // reg->mem followed by same  mem ->reg
4.  mov reg mem
     mov mem reg // mem->reg followed by same reg-> mem

SingleAccretion replied:
I wanted to note that the back to back moves can have the side effect changing the GC-ness of a location/register.

For example, in this case:
```
//         vmovupd  ymmword ptr[V01 rbp-50H], ymm0  # <-- last instruction
//         vmovupd  ymm0, ymmword ptr[V01 rbp-50H]  # <-- current instruction can be omitted.
```
(Let's pretend vectors can store GC refs, this obviously only applies to GPRs)

If `V01` was a non-GC variable, and `ymm0` in the first instruction held a non-GC value, but then the load was to a GC variable (`ymm0` would become GC-live), we cannot eliminate it.

Does the logic here handle this case already?

DeepakRajendrakumaran replied:
> Maybe we could call this `IsRedundantStackMov` or similar?
> 
> Someone on the JIT team may have a better suggestion or convention we could use here.

I like this. I think it's descriptive. The reason for keeping it R_S was trying to stick to existing convention but I agree oit might eb a little confusing

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -7055,12 +7055,170 @@ void emitter::emitIns_S(instruction ins, emitAttr attr, int varx, int offs)
+//                 ins  - The current instruction
+//                 fmt  - The current format
+//                 size - Operand size of current instruction
+//                 ireg - The current source/destination register
```
How does this differentiate `source` (store) vs `destination` (load)?

DeepakRajendrakumaran replied:
> How does this differentiate `source` (store) vs `destination` (load)?

Since both register and var can be either the source or destination, we can differentiate only using instruction format - IF_SWR_RRD vs IF_RWR_SRD. If SWR - we know it's writing to 'stack' and reading from register and vice versa. The logic to handle that is from line 7181

DeepakRajendrakumaran replied:
> Looks to be just based on `insFormat`.
> 
> Should we be asserting that the format is an expected one?

I'm skipping the optimization if format is not expected -

` if (!((emitLastIns->idInsFmt() == IF_SWR_RRD) || (emitLastIns->idInsFmt() == IF_RWR_SRD)))
    {
        return false;
    }`

tannergooding replied:
That's for the last ins format, not for the current instruction format.

I was thinking we likely should have a simple `assert((fmt == IF_SWR_RRD) || (fmt == IF_RWR_SRC))`

DeepakRajendrakumaran replied:
Ah. I see. Will do

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -7055,12 +7055,170 @@ void emitter::emitIns_S(instruction ins, emitAttr attr, int varx, int offs)
+            break;
+        }
+
+        case INS_movsdsse2:
+        case INS_movss:
+        {
+            // Clears the upper bits under VEX encoding
+            hasSideEffect = UseVEXEncoding();
+            break;
+        }
+
+        case INS_movsx:
+        case INS_movzx:
+        {
+            // Sign/Zero-extends the source
+            hasSideEffect = true;
+            break;
+        }
+
+#if defined(TARGET_AMD64)
+        case INS_movq:
+        {
+            // Clears the upper bits
+            hasSideEffect = true;
+            break;
+        }
+
+        case INS_movsxd:
+        {
+            // Sign-extends the source
+            hasSideEffect = true;
+            break;
+        }
+#endif // TARGET_AMD64
+
+        default:
+        {
+            unreached();
+        }
+    }
```
Most of this looks to be duplicate from https://github.com/dotnet/runtime/blob/9914c52c7cd4954a4205d5f66c562aa4dd5cce8b/src/coreclr/jit/emitxarch.cpp#L4389-L4390

Perhaps we can refactor it out to a helper method?

DeepakRajendrakumaran replied:
Yes, I agree. I'll go ahead and move it to a  helper. That should make it easier to handle teh upper bit issue with vex at 1 place if we come across a better solution later as well

DeepakRajendrakumaran committed:
Review changes:
1. Changing name to IsRedundantStackMov.
2. Add assert for format.
3. Move common code to helper.

DeepakRajendrakumaran committed:
Adding check to skip optimization when dealing with movs creating GC Live objs.

kunalspathak reviewed (approved):
I verified some of the asmdiffs and they looks good. Thank you for your contribution!
